crates/bevy_ecs/src/query/fetch.rs

```diff
@@ -810,11 +811,215 @@ unsafe impl<'a> WorldQuery for FilteredEntityMut<'a> {
     }
 }
 
+// The `State` that `EntityRefExcept` stores. This is simply the number of
+// registered components (so that the `Access` can be properly constructed) as
+// well as a list of excluded IDs.
+#[doc(hidden)]
+pub struct EntityRefExceptWorldQueryState<CIL>
+where
+    CIL: ComponentIdList,
+{
+    // The IDs of the excluded components.
+    ids: CIL,
+    // The total number of components that have been registered to the world.
+    component_count: usize,
+}
+
 /// SAFETY: access of `FilteredEntityRef` is a subset of `FilteredEntityMut`
 unsafe impl<'a> QueryData for FilteredEntityMut<'a> {
     type ReadOnly = FilteredEntityRef<'a>;
 }
 
+/// SAFETY: `EntityRefExcept` guards access to all components in the list `CL`
+/// and populates `Access` values so that queries that conflict with this access
+/// are rejected.
+unsafe impl<'a, CL> WorldQuery for EntityRefExcept<'a, CL>
+where
+    CL: ComponentList,
+{
+    type Fetch<'w> = UnsafeWorldCell<'w>;
+    type Item<'w> = EntityRefExcept<'w, CL>;
+    type State = EntityRefExceptWorldQueryState<CL::Ids>;
+
+    fn shrink<'wlong: 'wshort, 'wshort>(item: Self::Item<'wlong>) -> Self::Item<'wshort> {
+        item
+    }
+
+    fn shrink_fetch<'wlong: 'wshort, 'wshort>(fetch: Self::Fetch<'wlong>) -> Self::Fetch<'wshort> {
+        fetch
+    }
+
+    unsafe fn init_fetch<'w>(
+        world: UnsafeWorldCell<'w>,
+        _: &Self::State,
+        _: Tick,
+        _: Tick,
+    ) -> Self::Fetch<'w> {
+        world
+    }
+
+    const IS_DENSE: bool = true;
+
+    unsafe fn set_archetype<'w>(
+        _: &mut Self::Fetch<'w>,
+        _: &Self::State,
+        _: &'w Archetype,
+        _: &'w Table,
+    ) {
+    }
+
+    unsafe fn set_table<'w>(_: &mut Self::Fetch<'w>, _: &Self::State, _: &'w Table) {}
+
+    unsafe fn fetch<'w>(
+        world: &mut Self::Fetch<'w>,
+        entity: Entity,
+        _: TableRow,
+    ) -> Self::Item<'w> {
+        let cell = world.get_entity(entity).unwrap();
+        EntityRefExcept::new(cell)
+    }
+
+    fn update_component_access(
+        state: &Self::State,
+        filtered_access: &mut FilteredAccess<ComponentId>,
+    ) {
+        // `Access` specifies which components are allowed, not which are
+        // disallowed. So we iterate through all component IDs in the `World`,
+        // except for the excluded ones, and add them one-by-one to the
+        // `Access`. This is somewhat slow, but this code shouldn't be a hot
+        // path.
+        for component_index in 0..state.component_count {
+            let component_id = ComponentId::new(component_index);
+            if !state.ids.contains(component_id) {
+                filtered_access
+                    .access_mut()
+                    .add_component_read(component_id);
+            }
+        }
```

Comment:
This seems pretty suspicious. What happens when a component is initialized after `update_component_access` is called for this query?

Reply:
Yeah this seems to be unsound

```rs
    #[test]
    fn bad() {
        let mut world = World::new();
        world.init_component::<TestComponent>();

        let system = (
            QueryParamBuilder::new(|_| {}),
            QueryParamBuilder::new(|_| {}),
        )
            .build_state(&mut world)
            .build_system(
                |mut q1: Query<EntityMutExcept<TestComponent>>,
                 mut q2: Query<&mut TestComponent2>| {
                    let mut e1 = q1.single_mut();
                    let mut c1 = e1.get_mut::<TestComponent2>().unwrap();
                    let mut c2 = q2.single_mut();

                    assert!(!std::ptr::eq(&mut *c1, &mut *c2));
                },
            );

        world.spawn(TestComponent2(0));

        world.run_system_once(system);
    }
```

Reply:
The maximum component index is no longer relevant.